File: ts/test/debug-model.test.ts
/* Copyright (c) 2026 tabnas, MIT License */

// Composition test: the MultiSource grammar plugin layered with the official
// @tabnas/debug plugin. @tabnas/debug is a devDependency, but this resolves it
// dynamically and SKIPS when it is absent so the suite stays runnable outside
// the package; CI can point TABNAS_DEBUG_PATH at a sibling checkout's built
// plugin.

import { describe, it } from 'node:test'
import assert from 'node:assert'

import { Tabnas } from '@tabnas/parser'
import { jsonic } from '@tabnas/jsonic'
import { MultiSource } from '../dist/multisource'

function loadDebug(): any {
  const candidates = [process.env.TABNAS_DEBUG_PATH, '@tabnas/debug'].filter(
    Boolean,
  ) as string[]
  for (const c of candidates) {
    try {
      return require(c).Debug
    } catch {
      /* try next */
    }
  }
  return null
}

const Debug = loadDebug()
const skip = Debug ? false : '@tabnas/debug not available (set TABNAS_DEBUG_PATH)'

describe('compose: multisource + @tabnas/debug', () => {
  it('parses normally with the debug plugin installed', { skip }, () => {
    const tn = new Tabnas().use(jsonic).use(MultiSource, {})
    tn.use(Debug, { print: false, trace: false })
    assert.deepStrictEqual(
      JSON.parse(JSON.stringify(tn.parse('{"a":[1,2]}'))),
      { a: [1, 2] },
    )
  })

  it('debug.model() returns the structured grammar', { skip }, () => {
    const tn = new Tabnas().use(jsonic).use(MultiSource, {})
    tn.use(Debug, { print: false, trace: false })
    const m = tn.debug.model()

    // The structured rule set: the shared val/map/list/pair/elem rules plus
    // the MultiSource-specific `multisource` rule.
    assert.deepStrictEqual(
      m.rules.map((r: any) => r.name).sort(),
      ['elem', 'list', 'map', 'multisource', 'pair', 'val'],
    )

    // The entry rule and the installed plugins.
    assert.equal(m.config.start, 'val')
    assert.ok(
      m.plugins.some((p: any) => p.name === 'MultiSource'),
      'plugins should list MultiSource',
    )

    // Structural facts specific to this grammar: the `@`-marked multisource
    // directive wires `val` into a `multisource` rule, which in turn parses a
    // `val` for the resolved source.
    const val = m.rules.find((r: any) => r.name === 'val')
    assert.ok(
      val.open.some((a: any) => a.push === 'multisource'),
      'val should push multisource',
    )
    const multisource = m.rules.find((r: any) => r.name === 'multisource')
    assert.ok(
      multisource.open.some((a: any) => a.push === 'val'),
      'multisource should push val',
    )

    // The grammar portion is JSON-serialisable and round-trips.
    const grammar = {
      tokens: m.tokens,
      rules: m.rules,
      graph: m.graph,
      config: m.config,
      abnf: m.abnf,
    }
    assert.deepStrictEqual(
      JSON.parse(JSON.stringify(grammar)).rules,
      m.rules,
    )
  })
})

File: rs/tests/debug_model_test.rs
// Composition: the multisource plugin layered with the official
// tabnas-debug plugin, asserting the structured grammar model. Ports
// ts/test/debug-model.test.ts.
//
// The Rust port takes tabnas-debug as a dev-dependency on a sibling
// checkout rather than loading it dynamically, so this test never skips:
// a skip that nothing notices is how a composition stops being checked.

use tabnas_debug::{model, DebugOptions};
use tabnas_multisource::{multisource, MapResolver, MultiSourceOptions};

fn composed() -> tabnas::Tabnas {
    let mut parser = tabnas_jsonic::make();
    multisource(
        &mut parser,
        MultiSourceOptions::new(MapResolver::from([("a.jsonic", "a:1")])),
    )
    .expect("the MultiSource plugin installs");
    tabnas_debug::apply(&mut parser, DebugOptions::quiet()).expect("the Debug plugin installs");
    parser
}

#[test]
fn it_parses_normally_with_the_debug_plugin_installed() {
    let parser = composed();
    assert_eq!(
        parser.parse(r#"{"a":[1,2]}"#).expect("parses").to_string(),
        r#"{"a":[1,2]}"#
    );
    assert_eq!(
        parser
            .parse(r#"x:@"a.jsonic""#)
            .expect("parses")
            .to_string(),
        r#"{"x":{"a":1}}"#
    );
}

#[test]
fn the_model_reports_the_grammar() {
    let parser = composed();
    let model = model(&parser);

    // The rule set: the shared jsonic rules plus this plugin's own.
    let mut names: Vec<String> = model.rules.iter().map(|rule| rule.name.clone()).collect();
    names.sort();
    assert_eq!(names, ["elem", "list", "map", "multisource", "pair", "val"]);

    assert_eq!(model.config.start, "val");
    assert!(
        model
            .plugins
            .iter()
            .any(|plugin| plugin.name == "MultiSource"),
        "the plugins list names MultiSource: {:?}",
        model
            .plugins
            .iter()
            .map(|plugin| &plugin.name)
            .collect::<Vec<_>>()
    );

    // The structural facts specific to this grammar: the marked
    // directive wires `val` into a `multisource` rule, which in turn
    // parses a `val` for the resolved source.
    let val = model
        .rules
        .iter()
        .find(|rule| rule.name == "val")
        .expect("a val rule");
    assert!(
        val.open
            .iter()
            .any(|alt| alt.push.as_deref() == Some("multisource")),
        "val pushes multisource"
    );

    let rule = model
        .rules
        .iter()
        .find(|rule| rule.name == "multisource")
        .expect("a multisource rule");
    assert!(
        rule.open
            .iter()
            .any(|alt| alt.push.as_deref() == Some("val")),
        "multisource pushes val"
    );

    // The model is serializable and round-trips.
    let encoded = serde_json::to_string(&model).expect("the model serializes");
    let decoded: serde_json::Value = serde_json::from_str(&encoded).expect("and reads back");
    assert!(decoded.get("rules").is_some());
}
